src/libostree/ostree-repo-composefs.c

```diff
@@ -0,0 +1,631 @@
+      const guint8 *value_data = g_variant_get_fixed_array (value, &value_len, 1);
+
+      if (lcfs_node_set_xattr (node, (char *)name, (char *)value_data, value_len) != 0)
+        return glnx_throw_errno_prefix (error, "Setting composefs xattrs for %s", name);
+    }
+
+  return TRUE;
+}
+
+static gboolean
+checkout_one_composefs_file_at (OstreeRepo *repo, const char *checksum, struct lcfs_node_s *parent,
+                                const char *destination_name, GCancellable *cancellable,
+                                GError **error)
+{
+  g_autoptr (GInputStream) input = NULL;
+  g_autoptr (GVariant) xattrs = NULL;
+  struct lcfs_node_s *existing;
+
+  /* Validate this up front to prevent path traversal attacks */
+  if (!ot_util_filename_validate (destination_name, error))
+    return FALSE;
+
+  existing = lcfs_node_lookup_child (parent, destination_name);
+  if (existing != NULL)
+    return glnx_throw (error, "Target checkout file already exist");
+
+  g_autoptr (GFileInfo) source_info = NULL;
+  if (!ostree_repo_load_file (repo, checksum, &input, &source_info, &xattrs, cancellable, error))
+    return FALSE;
+
+  const guint32 source_mode = g_file_info_get_attribute_uint32 (source_info, "unix::mode");
+  const guint32 source_uid = g_file_info_get_attribute_uint32 (source_info, "unix::uid");
+  const guint32 source_gid = g_file_info_get_attribute_uint32 (source_info, "unix::gid");
+  const guint64 source_size = g_file_info_get_size (source_info);
+  const char *source_symlink_target = g_file_info_get_symlink_target (source_info);
+  const gboolean is_symlink
+      = (g_file_info_get_file_type (source_info) == G_FILE_TYPE_SYMBOLIC_LINK);
+
+  struct lcfs_node_s *node = lcfs_node_new ();
+  if (node == NULL)
+    return glnx_throw (error, "Out of memory");
+
+  /* Takes ownership on success */
+  if (lcfs_node_add_child (parent, node, destination_name) != 0)
+    {
+      lcfs_node_unref (node);
+      return glnx_throw_errno (error);
+    }
+
+  lcfs_node_set_mode (node, source_mode);
+  lcfs_node_set_uid (node, source_uid);
+  lcfs_node_set_gid (node, source_gid);
+  lcfs_node_set_size (node, source_size);
+  if (is_symlink)
+    {
+      if (lcfs_node_set_payload (node, source_symlink_target) != 0)
+        return glnx_throw_errno (error);
+    }
+  else if (source_size != 0)
+    {
+      char loose_path_buf[_OSTREE_LOOSE_PATH_MAX];
+      _ostree_loose_path (loose_path_buf, checksum, OSTREE_OBJECT_TYPE_FILE, OSTREE_REPO_MODE_BARE);
+      if (lcfs_node_set_payload (node, loose_path_buf) != 0)
+        return glnx_throw_errno (error);
+
+      guchar *known_digest = NULL;
+
+#ifdef HAVE_LINUX_FSVERITY_H
+      /* First try to get the digest directly from the bare repo file.
+       * This is the typical case when we're pulled into the target
+       * system repo with verity on and are recreating the composefs
+       * image during deploy. */
+      char buf[sizeof (struct fsverity_digest) + OSTREE_SHA256_DIGEST_LEN];
```

Comment:
Maybe a best practice to use `FS_VERITY_MAX_DIGEST_SIZE` here instead of `OSTREE_SHA256_DIGEST_LEN`?

(We could also add a static assertion that the latter is less than the former)